pkg/handler/v2/provider/kubevirt.go

```diff
@@ -42,60 +42,106 @@ type KubeVirtGenericReq struct {
 }
 
 // KubeVirtGenericNoCredentialReq represent a generic KubeVirt request with cluster credentials.
-// swagger:parameters listKubeVirtVMIPresetsNoCredentials listKubevirtStorageClassesNoCredentials
+// swagger:parameters listKubeVirtVMIPresetsNoCredentials listKubevirtStorageClassesNoCredentials listKubeVirtInstancetypesNoCredentials listKubeVirtPreferencesNoCredentials
 type KubeVirtGenericNoCredentialReq struct {
 	cluster.GetClusterReq
 }
 
+func kubeconfigFromRequest(ctx context.Context, request interface{}, presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter) (string, error) {
+	req, ok := request.(KubeVirtGenericReq)
+	if !ok {
+		return "", fmt.Errorf("incorrect type of request, expected = KubeVirtGenericReq, got %T", request)
+	}
+	kubeconfig := req.Kubeconfig
+
+	userInfo, err := userInfoGetter(ctx, "")
+	if err != nil {
+		return "", common.KubernetesErrorToHTTPError(err)
+	}
+	if len(req.Credential) > 0 {
+		preset, err := presetsProvider.GetPreset(ctx, userInfo, req.Credential)
+		if err != nil {
+			return "", utilerrors.New(http.StatusInternalServerError, fmt.Sprintf("can not get preset %s for user %s", req.Credential, userInfo.Email))
+		}
+		if credentials := preset.Spec.Kubevirt; credentials != nil {
+			kubeconfig = credentials.Kubeconfig
+		}
+	}
+	return kubeconfig, nil
+}
+
 // KubeVirtVMIPresetsEndpoint handles the request to list available KubeVirtVMIPresets (provided credentials).
 func KubeVirtVMIPresetsEndpoint(presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter, settingsProvider provider.SettingsProvider) endpoint.Endpoint {
 	return func(ctx context.Context, request interface{}) (interface{}, error) {
-		req := request.(KubeVirtGenericReq)
-		kubeconfig := req.Kubeconfig
+		kubeconfig, err := kubeconfigFromRequest(ctx, request, presetsProvider, userInfoGetter)
```

Comment:
@lsviben what's the difference between requests with `WithClusterCredentialsEndpoint` sufix and without?? I see in both cases we have access to credentials. I am trying to understand the reason why we have those two different endpoints.

Reply:
From my understanding, @lsviben to confirm.
1) Without the `WithClusterCredentialsEndpoint` is used in the cluster creation flow, before the cluster and credentials secrets are created.
Request header contains the credentials: the KubeVirt kubeconfig or the Kubermatic preset containing the kubeconfig.
2) With the WithClusterCredentialsEndpoint is used once the Cluster is created and credentials stored in the appropriate secret in the `kubermatic` namespace.
`/api/v2/projects/{project_id}/clusters/{cluster_id}/providers/kubevirt/instancetypes`, credentials are not passed in the request, but retrieved from the secret.

Reply:
@hdurand0710 is correct :)

Reply:
thanks for detailed explanation